### DISCUSSION TITLE
Publicize ASC sources?

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2156] [Q&A]
Can Harman publicize its ActionScript compiler? Flex SDK and Apache Royale did the same... If there are authority rights, then nevermind... Thanks!

**Comment by ajwfrost** [comment]
Not under our current license agreement with Adobe (as it comes under the generic "source code" rules), but I'll discuss with them; there are quite a lot of similarities between what we have and the Royale compiler, and it's already been pointed out a couple of times to us that it would be great to have these changes also going into those ones..

**Reply by henke37** [reply]
When asking Adobe, see if you can get the Animate "publishing" sub module too. It'd be great to not require a gui tool for building stuff.

**Reply by pnico** [reply]
It seems like this should be addressed before adding new AS3 language features. It doesn't make sense for there to be multiple open-source versions of avmplus and ASC floating around and then have Harman using its own proprietary compiler and runtime that are incompatible with community contributions like the VSCode plugin. Do the runtime and compiler actually touch anything that would prevent them from being able to be open-sourced, other than whatever the license agreement says about it?

**Reply by ajwfrost** [reply]
@pnico I did discuss this ... yes there are, I think primarily it's the font stuff that's proprietary. I was discussing this with one of the engineers who had worked on the earlier stuff and the Apache handovers, it may be that the best option is to take maybe the Royale compiler and try to have that as the primary one, but that would then also require us to be contributing to open source code and there are extra hoops and issues we'd have to jump through for that. But I may be able to think of a workaround here... I'm also just thinking, the proprietary font code may be stuff that we don't actually have ourselves in source format, given an error I was looking at recently. So, I would agree with you, this would be a better route forwards, and if we're ultimately able to create a version that supports the Flex MXML features as well then it would mean a simpler maintenance situation all round...

@henke37 I've only just seen your comment, sorry.. what do you mean by this? You shouldn't need a GUI for publishing, or rather, for packaging; Animate should always be calling down into other tools that are provided either within it or from the AIR SDK, so it should be possible to replicate what it does but call things from the command line...

**Reply by henke37** [reply]
Let me know how to publish a fla file without running `animate.exe`. JSFL can automate it, but that's not a scalable solution.

**Reply by ajwfrost** [reply]
In terms of "publishing", I guess we should clarify that there are two steps:
1) creating the SWF file - this is what Animate does internally to turn all your assets, symbols, timelines etc into a SWF file
2) packaging the application - this is the part where it calls the AIR SDK to package the SWF along with other resources into a deployable application - e.g. an APK file for Android, or IPA file for iOS, etc.

This second step is what I was referring to as this will call down to ADT and it would be possible to find out how this is being called and then just do the same thing from the command line. (Try adding/uncommenting `DebugOut=true` from the adt.cfg file in your AIR SDK's 'lib' folder, then check for a file `c:\users\username\adt.log`)

But the first bit is something that needs Animate, afaik.

**Reply by henke37** [reply]
You separate compiling and packaging. Animate does not. Point is, can't compile a fla file without `Animate.exe`, which is what I'd like remedied.

**Reply by marchbold** [reply]
@henke37 The way we do it is we use Animate only to develop assets and export them to a swc. Then we use that swc as the visual assets source for our application code that we use an ide and the air sdk directly to compile and package. We find this approach is much better suited to automation.

**Reply by henke37** [reply]
That doesn't fly for me, since I tend to heavily integrate symbols and classes. Real classes, not dummy classes that only serve as names for code to spawn. I'm talking meaningful instance names.

**Reply by ajwfrost** [reply]
@henke37 can I just check, what you're after is basically a command-line driven version of Animate so that things can be rebuilt (including symbols and separate AS3 files) based on changes in the AS3 etc? As far as I'm aware this would need to be triggered by JSFL, it's years since I've used that in a production project so not sure if a script can be kicked off automatically on boot..

But the general use of symbols, timelines etc is what Animate is there for, with the information stored in their proprietary .fla format, so generating those assets into the SWF tags is the bit of magic that's part of what you pay for :-)

**Reply by marchbold** [reply]
@henke37 Why can't you do that using the export to swc approach? Everything defined in the MC is still available as long as it's marked for export etc. You can inherit a class in your app code from the visual asset in animate and still obtain tight integration. Control the timeline if you need etc.

**Reply by henke37** [reply]
Yup, it can be done with jsfl. But it's not as nice as a proper batch program that doesn't show ui.

As for the payment argument, I say that you pay for the nice animation tool and what little IDE is provided. The compiler itself is a freebie in my book.
